cosmwasm/contracts/ibc-translator/src/execute.rs

```diff
@@ -0,0 +1,310 @@
+use cosmwasm_std::{
+    to_binary, Binary, Coin, CosmosMsg, Deps, DepsMut, Empty, Env, Event, MessageInfo,
+    QueryRequest, Response, SubMsg, Uint128, WasmMsg, WasmQuery,
+};
+use cw_token_bridge::msg::{
+    Asset, AssetInfo, ExecuteMsg as TokenBridgeExecuteMsg, QueryMsg as TokenBridgeQueryMsg,
+    TransferInfoResponse,
+};
+use cw_wormhole::byte_utils::ByteUtils;
+
+use cw20_wrapped_2::msg::ExecuteMsg as Cw20WrappedExecuteMsg;
+use serde_wormhole::RawMessage;
+use std::str;
+use wormhole_bindings::{
+    tokenfactory::{TokenFactoryMsg, TokenMsg},
+    WormholeQuery,
+};
+use wormhole_sdk::{
+    ibc_translator::{Action, GovernancePacket},
+    vaa::{Body, Header},
+    Chain,
+};
+
+use crate::{
+    msg::COMPLETE_TRANSFER_REPLY_ID,
+    state::{
+        CHAIN_TO_CHANNEL_MAP, CURRENT_TRANSFER, CW_DENOMS, TOKEN_BRIDGE_CONTRACT, VAA_ARCHIVE,
+    },
+};
+
+pub enum TransferType {
+    Simple { fee: Uint128 },
+    ContractControlled { payload: Binary },
+}
+
+/// Calls into the wormhole token bridge to complete the payload3 transfer.
+pub fn complete_transfer_and_convert(
+    deps: DepsMut<WormholeQuery>,
+    env: Env,
+    info: MessageInfo,
+    vaa: Binary,
+) -> Result<Response<TokenFactoryMsg>, anyhow::Error> {
+    // get the token bridge contract address from storage
+    let token_bridge_contract = TOKEN_BRIDGE_CONTRACT
+        .load(deps.storage)
+        .context("could not load token bridge contract address")?;
+
+    // craft the token bridge execute message
+    // this will be added as a submessage to the response
+    let token_bridge_execute_msg = to_binary(&TokenBridgeExecuteMsg::CompleteTransferWithPayload {
+        data: vaa.clone(),
+        relayer: info.sender.to_string(),
+    })
+    .context("could not serialize token bridge execute msg")?;
+
+    let sub_msg = SubMsg::reply_on_success(
+        CosmosMsg::Wasm(WasmMsg::Execute {
+            contract_addr: token_bridge_contract.clone(),
+            msg: token_bridge_execute_msg,
+            funds: vec![],
+        }),
+        COMPLETE_TRANSFER_REPLY_ID,
+    );
+
+    // craft the token bridge query message to parse the payload3 vaa
+    let token_bridge_query_msg = to_binary(&TokenBridgeQueryMsg::TransferInfo { vaa })
+        .context("could not serialize token bridge transfer_info query msg")?;
+
+    let transfer_info: TransferInfoResponse = deps
+        .querier
+        .query(&QueryRequest::Wasm(WasmQuery::Smart {
+            contract_addr: token_bridge_contract,
+            msg: token_bridge_query_msg,
+        }))
+        .context("could not parse token bridge payload3 vaa")?;
+
+    // DEFENSE IN-DEPTH CHECK FOR PAYLOAD3 VAAs
+    // ensure that the transfer vaa recipient is this contract.
+    // we should never process any VAAs that are not directed to this contract.
+    let target_address = (&transfer_info.recipient.as_slice()).get_address(0);
+    let recipient = deps.api.addr_humanize(&target_address)?;
+    ensure!(
```

Comment:
Is there a vulnerability in the existing token bridge contract that would allow someone else to redeem this transfer with payload?

Reply:
I don't think so, this check here should restrict that:
https://github.com/wormhole-foundation/wormhole/blob/main/cosmwasm/contracts/token-bridge/src/contract.rs#L838

Reply:
We had added this to the sei-token-translator contract (which this contract was adapted from) as part of the internal security review for a defense in depth check.
https://github.com/everstake/sei-token-translator/tree/main